### PR TITLE
Feat/auth cookie update

### LINKED ISSUE #206: Have `setAuthCookie` return void or an error instead of responding to the http request
# Feature request

## Is your feature request related to a problem? Please describe.

The `auth.api.setAuthCookie` method seems to only work for `Next.js` and `Express` as stated in the comment [here](https://github.com/supabase/gotrue-js/blob/master/src/GoTrueApi.ts#L376-L380). 

When using other web frameworks, there are a few issues. For example, `Fastify` web framework doesn't have a `res.setHeader` method so [this](https://github.com/supabase/gotrue-js/blob/d7b334a4283027c65814aa81715ffead262f0bfa/src/lib/cookies.ts#L155) fails. Or if one is to use `Node.js` http server, `res.status` fails [here](https://github.com/supabase/gotrue-js/blob/master/src/GoTrueApi.ts#L400)

## Describe the solution you'd like

A solution that I see would work is to not return the http request within the `auth.api.setAuthCookie` method. Rather have it return an error or void and let the developer handle returning the http request. 

This would allow `setAuthCookie` to do its cookie setting/deleting and also allow any web framework to use this method with no issues.

### PR DESCRIPTION
**Pull request by thorwebdev**
- [x] Updates `setAuthCookie(req, res)` to set `sb-access-token` and `sb-refresh-token` cookies.
- [x] Updates `getUserByCookie(req, res?)` to automatically try and refresh the access token if expired. Note: this requires `res` to be passed!
- [x] Introduces `getAuthCookieString(req, res): string[]` to make it easier to set auth cookies in frameworks that aren't supported by `setAuthCookie`. Closes #206

### CODE REVIEW

**Comment by dthyresson** [comment]
> anything that stands out to you?

@thorwebdev What do the response headers look like?

When you say request, what is making the request and what is receiving the request?

**Comment by thorwebdev** [comment]
Gotrue is making the request via http.Redirect in the callback route. 

Next.js api route is receiving the request.

**Comment by dthyresson** [comment]
> Gotrue is making the request via http.Redirect in the callback route.
> 
> Next.js api route is receiving the request.

When in doubt .... CORS?

Also, see `Preflight Requests Block Cookies`: https://medium.com/swlh/7-keys-to-the-mystery-of-a-missing-cookie-fdf22b012f09 


With RedwoodJS's cookie dbAuth provider, ran into a case where CORS needed to be set https://github.com/redwoodjs/redwood/pull/4150

**Comment by bnjmnt4n** [comment]
Based on my understanding of cookies, this approach doesn't work since the cookies returned by GoTrue will be associated with the GoTrue domain. GoTrue redirects to the given API route which is on a different domain. The browser will not send the cookies associated with the GoTrue domain to any other domain since that would trivially lead to exploitation.

**Comment by thorwebdev** [comment]
@bnjmnt4n yes, you are correct, this approach only works if gotrue is running on the same domain.

**Review by kangmingtay — CHANGES_REQUESTED** [review]

**Comment by thorwebdev** [comment]
@dthyresson would you mind testing the `getAuthCookieString` approach with your Fastify app?

File: src/GoTrueApi.ts [hunk 1]
```diff
@@ -513,15 +526,41 @@ export default class GoTrueApi {
...
+              domain: this.cookieOptions.domain,
+              maxAge: this.cookieOptions.lifetime ?? 0,
+              path: this.cookieOptions.path,
+              sameSite: this.cookieOptions.sameSite,
+            }))
+          )
+          return { token: data.access_token, user: data.user, data: data.user, error: null }
+        }
+      }
+      return { token: access_token, user, data: user, error: null }
```
[the remark this answers is not in the document]
Comment (thorwebdev) [reply src/GoTrueApi.ts]: For you most certainly ;)

**Comment by alaister** [comment]
Hey,
I'm missing a lot of context here, but I'm keen to see token refreshing for SSR type web apps.

What happens when the cookie tokens and the client side tokens become out of sync? For example when the token gets refreshed server side, and then the client also tries to refresh after the page loads.
Does the GoTrue backend allow the same refresh token to be used multiple times within a short period of time, making this a non-issue?

I spun up this PR locally, logged in, and then waited for my session to expire on the `/profile` route. After refreshing, the token refreshed server side no problem, but upon navigating back to the `/` route, it seemed I was logged out client side.

Please let me know if I can help out!

**Comment by thorwebdev** [comment]
Yes, very good point @alaister, you will need to decide what your source of truth is. E.g. here you will always need to fetch the user from the server, e.g. via https://github.com/supabase/gotrue-js/blob/feat/auth-cookie-update/example/next-auth0-migration-example/pages/api/me.js

In the future we're planning to move fully away from using localStorage, but there are a couple things that need to happen first.

**Review by kangmingtay — APPROVED** [review]

**Pull request merged**